### PR TITLE
added transfer learning support for symbols embedding layer

### PR DESCRIPTION
**Pull request by CookiePPP**

### CODE REVIEW

**Comment by CookiePPP** [comment]
image

**Comment by CookiePPP** [comment]
image

**Comment by CookiePPP** [comment]
image

**Comment by CookiePPP** [comment]
image

**Comment by CookiePPP** [comment]
image

**Comment by CookiePPP** [comment]
Summary:
With `batch_size = 27`;
| Learning Rate (per 64) | Val Loss | Attention Strength |
| ------------- | ------------- | ------------- |
| 1.0e-4 | 0.3325 | 0.52 |
| 2.0e-4 | 0.3151 | 0.49 |
| 4.0e-4 | 0.3103 | 0.51 |
| 1.5e-3 | 0.3521 | 0.51 |


old
| Learning Rate (per 64) | Val Loss | Attention Strength |
| ------------- | ------------- | ------------- |
| 4.0e-4 | 0.4355 | 0.20 |

**Comment by BenAAndrew** [comment]
> Not sure how much familiarity you have with Machine Learning, but to be clear;

Yep, I was aware of all of that but a very worthwhile point to mention. I have considered making transfer learning non-optional in the past. I think this concept is still confusing a lot of new users so if you can find a better way of describing its importance to noobs than what is written in the app I'd be more than happy to change: https://github.com/BenAAndrew/Voice-Cloning-App/blob/main/application/static/train.html#L41

File: training/checkpoint.py [hunk 1]
```diff
@@ -1,6 +1,16 @@
 import os
 import torch
-
+from typing import Optional
+from unidecode import unidecode
+
+NVIDIA_ALPHABET = ['_', '-', '!', "'", '(', ')', ',', '.', ':', ';', '?', ' ',
```
Comment (BenAAndrew) [inline training/checkpoint.py:6]: We already have a default alphabet defined at https://github.com/BenAAndrew/Voice-Cloning-App/blob/main/training/train.py#L35. Also, will this cause problems for non-English datasets or datasets that do not contain cmudict?

Comment (BenAAndrew) [inline training/checkpoint.py:6]: We already have a default alphabet defined at https://github.com/BenAAndrew/Voice-Cloning-App/blob/main/training/train.py#L35. Also, will this cause problems for non-English datasets or datasets that do not contain cmudict?

File: training/checkpoint.py [hunk 2]
```diff
@@ -36,7 +46,42 @@ def load_checkpoint(checkpoint_path, model, optimizer, train_loader):
...
+        
+        # transfers non-ascii symbols from pretrained model to new model  e.g: 'e' -> 'é'
+        elif unidecode(new_symbol) in weight_dict:
+            embedding_layer.weight.data[symbol_index] = weight_dict[unidecode(new_symbol)]
+        
+        # transfers upper-case symbols from pretrained model to new model  e.g: 'E' -> 'e'
+        elif new_symbol.upper() in weight_dict:
+            embedding_layer.weight.data[symbol_index] = weight_dict[new_symbol.upper()]
+        
+        # transfers lower-case symbols from pretrained model to new model  e.g: 'e' -> 'E'
```
Comment (BenAAndrew) [inline training/checkpoint.py:75]: Do we need to convert to uppercase?

File: training/checkpoint.py [hunk 3]
```diff
@@ -36,7 +46,42 @@ def load_checkpoint(checkpoint_path, model, optimizer, train_loader):
     return model, optimizer, iteration, epoch
 
 
-def warm_start_model(checkpoint_path, model, ignore_layers=["embedding.weight"]):
+def transfer_symbols_embedding(original_embedding_weight: torch.Tensor, embedding_layer, original_symbols: Optional[list], new_symbols: list):
```
Comment (BenAAndrew) [inline training/checkpoint.py:49]: Please add a docstring for this function

File: training/checkpoint.py [hunk 4]
```diff
@@ -49,8 +94,12 @@ def warm_start_model(checkpoint_path, model, ignore_layers=["embedding.weight"])
     model : Tacotron2
         tacotron2 model to load checkpoint into
     ignore_layers : list (optional)
-        list of layers to ignore (default is ['embedding.weight'])
-
+        list of layers to ignore (default is ["embedding.weight"])
+    symbols : list
+        list of text symbols used by the fresh model currently loaded
+    old_symbols : list (optional)
```
Comment (BenAAndrew) [inline training/checkpoint.py:100]: I don't think we should pass this as a parameter. Can we not get it from the existing checkpoint each time?

File: training/train.py [hunk 5]
```diff
@@ -232,6 +232,16 @@ def train(
     logging.info("Saving model and optimizer state at iteration {} to {}".format(iteration, checkpoint_path))
 
 
+def calc_avgmax_attention(batch, y_pred):
```
Comment (BenAAndrew) [inline training/train.py:235]: Please add a docstring

File: training/training.md [hunk 6]
```diff
@@ -11,4 +11,9 @@ Once we have a dataset that has been preprocessed, we can begin training the voi
 - batch_size (optional): Batch size/ memory usage. Calculated automatically if not given
 
 ## How to run
-`python train.py -m dataset/metadata.csv -d dataset/wavs -o dataset/checkpoints `
+
+1. Download NVIDIA's [Tacotron 2](https://github.com/NVIDIA/tacotron2#training-using-a-pre-trained-model) model
+   
+2. Move into `tacotron2_statedict.pt` into `training` folder
+
+3. Run `python train.py -m dataset/metadata.csv -d dataset/wavs -o dataset/checkpoints -t tacotron2_statedict.pt`
```
Comment (BenAAndrew) [inline training/training.md:19]: Nice 👍

File: training/utils.py [hunk 7]
```diff
@@ -8,7 +8,8 @@
 
 CHECKPOINT_SIZE_MB = 333
 BATCH_SIZE_PER_GB = 2.5
-LEARNING_RATE_PER_BATCH = 3.125e-5
+LEARNING_RATE_PER_64 = 4e-4 # corrected learning rate to match paper. # Tacotron2 alignment forms best between 2e-4 and 5e-5
```
Comment (BenAAndrew) [inline training/utils.py:11]: Please remove second # in comment

**Review by BenAAndrew — CHANGES_REQUESTED** [review]
Awesome PR! Only a few minor comments to make it easier to maintain

File: training/checkpoint.py (hunk 1 above)
Comment (CookiePPP) [reply training/checkpoint.py:6]: When transferring embedding layers with different symbols, you need to know the original symbols and the new symbols since the index of the symbol is it's location in the layer.
NVIDIA_ALPHABET contains the alphabet in the exact order used by nvidia's pretrained model on their repo.

File: training/checkpoint.py (hunk 2 above)
Comment (CookiePPP) [reply training/checkpoint.py:75]: I included all conversion cases I could think of just to make sure it always works.

File: training/checkpoint.py (hunk 4 above)
Comment (CookiePPP) [reply training/checkpoint.py:100]: I can make that modification for new checkpoints, but not nvidia's as they do not save that information in the model file

File: training/checkpoint.py (hunk 1 above)
Comment (BenAAndrew) [reply training/checkpoint.py:6]: OK Cool, that makes sense. Are we confident this won't cause problems for non-English datasets or datasets that do not contain cmudict?

File: training/checkpoint.py (hunk 4 above)
Comment (BenAAndrew) [reply training/checkpoint.py:100]: Let's make the assumption that we can either use the NVIDIA default or get it from the checkpoint for now if that's OK

File: training/utils.py (hunk 7 above)
Comment (CookiePPP) [reply training/utils.py:11]: little busy, do it for me if you can, overwise wait for me

File: training/train.py (hunk 5 above)
Comment (CookiePPP) [reply training/train.py:235]: little busy, do it for me if you can, overwise wait for me

File: training/checkpoint.py (hunk 3 above)
Comment (CookiePPP) [reply training/checkpoint.py:49]: little busy, do it for me if you can, overwise wait for me

Comment (BenAAndrew) [reply training/checkpoint.py:49]: I'll give it a go, we'll also need unit tests

File: training/checkpoint.py (hunk 1 above)
Comment (CookiePPP) [reply training/checkpoint.py:6]: It will not cause problems. You can't initialize the embedding layer worse than scratch.

https://github.com/pytorch/pytorch/blob/1901c675e1cf9626a25d96bcc64b4952c64d56a2/torch/nn/modules/sparse.py#L148-L150
If you do not use this code then it will use literally random noise. This code cannot make the model worse than filling the embedding with random noise haha.

Comment (CookiePPP) [reply training/checkpoint.py:6]: https://github.com/CookiePPP/cookietts/tree/experimental/CookieTTS

Also to be clear, I've used similar code for my modules for over a year now. It works.

Comment (BenAAndrew) [reply training/checkpoint.py:6]: Awesome! Just checking that it won't impact other use cases (like other languages) 😄

File: training/checkpoint.py (hunk 3 above)
Comment (BenAAndrew) [reply training/checkpoint.py:49]: Suggestion: 

"""
Transfer embedding information from transfer learning model to reduce embedding time.
If symbol is not found it is initialised with mean/std.

Parameters
----------
original_embedding_weight : Torch.tensor
    Checkpoint embeddings
embedding_layer : ???
    Model embedding layer
original_symbols : list (optional)
    list of current symbols (defaults to NVIDIA_ALPHABET)
new_symbols : list
    list of text symbols used by the fresh model currently loaded
"""

File: training/train.py (hunk 5 above)
Comment (BenAAndrew) [reply training/train.py:235]: I'm not sure I entirely understand the process so I'll leave this to you if that's OK

**Comment by BenAAndrew** [comment]
We need to add/update unit tests, I'll start looking into this now

File: training/train.py [hunk 8]
```diff
@@ -177,6 +176,7 @@ def train(
             y, y_pred = process_batch(batch, model)
 
             loss = criterion(y_pred, y)
+            avg_prob = calc_avgmax_attention(batch[-1], batch[1], y_pred[-1])
```
Comment (BenAAndrew) [inline training/train.py:179]: Could we move this to line 197 and display in that message so that it is done every time we validate rather than every iteration

**Comment by BenAAndrew** [comment]
Making changes in other PR
